src/realm/list.cpp

```diff
@@ -564,6 +584,99 @@ void Lst<ObjKey>::clear()
     }
 }
 
+template <>
+void Lst<ObjLink>::do_set(size_t ndx, ObjLink target_link)
+{
+    ObjLink old_link = get(ndx);
+    CascadeState state(old_link.get_obj_key().is_unresolved() ? CascadeState::Mode::All : CascadeState::Mode::Strong);
+    bool recurse = m_obj.replace_backlink(m_col_key, old_link, target_link, state);
+
+    m_tree->set(ndx, target_link);
+
+    if (recurse) {
+        auto origin_table = m_obj.get_table();
+        _impl::TableFriend::remove_recursive(*origin_table, state); // Throws
+    }
+}
+
+template <>
+void Lst<ObjLink>::do_insert(size_t ndx, ObjLink target_link)
+{
+    m_obj.set_backlink(m_col_key, target_link);
+    m_tree->insert(ndx, target_link);
+}
+
+template <>
+void Lst<ObjLink>::do_remove(size_t ndx)
+{
+    ObjLink old_link = get(ndx);
+    CascadeState state(old_link.get_obj_key().is_unresolved() ? CascadeState::Mode::All : CascadeState::Mode::Strong);
+
+    bool recurse = m_obj.remove_backlink(m_col_key, old_link, state);
+
+    m_tree->erase(ndx);
+
+    if (recurse) {
+        auto table = m_obj.get_table();
+        _impl::TableFriend::remove_recursive(*table, state); // Throws
+    }
+}
+
+template <>
+void Lst<Mixed>::do_set(size_t ndx, Mixed value)
+{
+    ObjLink old_link;
+    ObjLink target_link;
+    Mixed old_value = get(ndx);
+
+    if (old_value.get_type() == type_TypedLink) {
+        old_link = old_value.get<ObjLink>();
+    }
+    if (value.get_type() == type_TypedLink) {
+        target_link = value.get<ObjLink>();
+    }
+
+    CascadeState state(old_link.get_obj_key().is_unresolved() ? CascadeState::Mode::All : CascadeState::Mode::Strong);
+    bool recurse = m_obj.replace_backlink(m_col_key, old_link, target_link, state);
+
+    m_tree->set(ndx, value);
+
+    if (recurse) {
+        auto origin_table = m_obj.get_table();
+        _impl::TableFriend::remove_recursive(*origin_table, state); // Throws
+    }
+}
+
+template <>
+void Lst<Mixed>::do_insert(size_t ndx, Mixed value)
+{
+    if (value.get_type() == type_TypedLink) {
+        m_obj.set_backlink(m_col_key, value.get<ObjLink>());
+    }
+    m_tree->insert(ndx, value);
+}
+
+template <>
+void Lst<Mixed>::do_remove(size_t ndx)
+{
+    ObjLink old_link;
+    Mixed old_value = get(ndx);
+
+    if (old_value.get_type() == type_TypedLink) {
+        old_link = old_value.get<ObjLink>();
+    }
+
+    CascadeState state(old_link.get_obj_key().is_unresolved() ? CascadeState::Mode::All : CascadeState::Mode::Strong);
+    bool recurse = m_obj.remove_backlink(m_col_key, old_link, state);
```

Comment:
Is this correct even if above condition is false and old_link is default constructed.

Reply:
It is correct, as `remove_backlink` is a null operation if `old_link` is false. I will consider a more readable approach